spec.bs

```diff
@@ -426,17 +427,74 @@ To <dfn>start bidding and scoring</dfn> given an [=auction config=] |auctionConf
           1. [=map/Set=] |perSignalsUrlGenerator|[|joiningOrigin|] to « |ig| ».
         1. Otherwise:
           1. [=list/Append=] |ig| to |perSignalsUrlGenerator|[|joiningOrigin|].
-1. [=map/For each=] |buyer| → |perBuyerGenerator| of |bidGenerators|:
-  1. [=map/For each=] |signalsUrl| -> |perSignalsUrlGenerator| of |perBuyerGenerator|:
-    1. TODO: Fetch |signalsUrl| synchronously.
-    1. [=map/For each=] |joiningOrigin| -> |groups| of |perSignalsUrlGenerator|:
-      1. [=list/For each=] |ig| of |groups|: 
-        1. Let |biddingScript| be the result of [=fetch bidding or scoring javascript=] with |ig|'s
-          [=interest group/bidding url=].
-        1. If |biddingScript| is an error, [=iteration/continue=].
-        1. TODO: [=Evaluate script=] with [=evaluate script/script=] set to |biddingScript|,
-          [=evaluate script/functionName=] set to "generateBid", and [=evaluate script/arguments=]
-          set to generateBid's arguemnts.
+1. Return |bidGenerators|.
+
+</div>
+
+<div algorithm="generate and score bids">
+
+To <dfn>generate and score bids</dfn> given an [=auction config=] |auctionConfig|, an
+[=ordered map=] |bidGenerators|, a [=global object=] |global|, and a [=promise type=] |p|:
+1. Let |topScore|, and |numTopBids| be 0.
+1. Let |winningAdUrl| be null.
+1. Let |allBidsGeneratedAndScored| be null.
+1. Let |queue| be the result of [=starting a new parallel queue=].
+1. [=parallel queue/enqueue steps|Enqueue the following steps=] to |queue|:
+  1. If |auctionConfig|'s [=auction config/component auctions=] are [=list/is empty|empty=]:
+    1. TODO: Create a seller script runner.
+  1. Otherwise:
+    1. [=list/For each=] |component| in |auctionConfig|'s [=auction config/component auctions=]:
+      1. [=generate and score bids=] with |component|.

```

Comment:
This seems wrong right? We're not passing in all of the required arguments. Should we make this algorithm recursive compatible or something, or delegate pieces of it out to a different algorithm? Making this algorithm compatible with recursive calls might not be too bad we just have to make sure we don't try and resolve the promise N different times in N different queued tasks... Also I'm not sure that we want to create _a new_ parallel queue for each time.. that seems a little explosive.

Reply:
@domfarolino +1 to Paul's question about the parallization. Currently the component auctions and separate bidders both run synchronously one after another. Is there any suggestion about whether we want to inform people that bidders can run in parallel (component auctions as well)?

Reply:
Done. PTAL. Thanks!

Reply:
Yeah if the intention is indeed to make all of those routines run in parallel _from each other_ then that sounds good. I wasn't too familiar with the intended flow here so I just wanted to make sure that we were't accidentally blowing things up. Thanks. And I think it's good that we moved the promise resolution stuff out of "generate and score bids", that looks clean.

Reply:
added a queue to run those routines in parallel.